File: bouyomi_client.py
import socket
import struct
import urllib.request
import urllib.parse

class BouyomiClient:
    """棒読みちゃんと通信するクライアント"""
    
    def __init__(self, host='localhost', port=50080, http_port=50080):
        self.host = host
        self.port = port
        self.http_port = http_port
    
    def speak(self, message, speed=-1, tone=-1, volume=-1, voice=0):
        """
        棒読みちゃんに読み上げを依頼
        
        Args:
            message: 読み上げるメッセージ
            speed: 速度 (-1: デフォルト)
            tone: 音程 (-1: デフォルト)
            volume: 音量 (-1: デフォルト)
            voice: 声質 (0: デフォルト)
        """
        try:
            # まずHTTP通信を試す
            if self.speak_http(message, speed, tone, volume, voice):
                return True
            
            # HTTP通信が失敗したらTCP/IP通信を試す
            return self.speak_tcp(message, speed, tone, volume, voice)
        except Exception as e:
            print(f"棒読みちゃんへの送信に失敗: {e}")
            return False
    
    def speak_http(self, message, speed=-1, tone=-1, volume=-1, voice=0):
        """HTTP通信で棒読みちゃんに送信"""
        try:
            # URLパラメータを構築
            params = {
                'text': message,
                'voice': voice,
                'volume': volume,
                'speed': speed,
                'tone': tone
            }
            
            # 不要なパラメータを削除
            params = {k: v for k, v in params.items() if v != -1}
            
            # URLを構築
            query = urllib.parse.urlencode(params)
            url = f"http://{self.host}:{self.http_port}/talk?{query}"
            
            print(f"HTTP通信で送信: {url}")
            
            # HTTP通信
            with urllib.request.urlopen(url, timeout=5) as response:
                print(f"HTTP応答: {response.status}")
            
            return True
        except Exception as e:
            print(f"HTTP通信での送信に失敗: {e}")
            return False
    
    def speak_tcp(self, message, speed=-1, tone=-1, volume=-1, voice=0):
        """TCP/IP通信で棒読みちゃんに送信"""
        try:
            # メッセージをUTF-8でエンコード
            message_bytes = message.encode('utf-8')
            
            # パケットを構築
            packet = struct.pack(
                '<HhhhhBi',
                0x0001,  # Command
                speed,
                tone,
                volume,
                voice,
                0,  # Encoding (UTF-8)
                len(message_bytes)
            ) + message_bytes
            
            # ソケット通信
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.settimeout(5)
                print(f"TCP/IP通信で送信: {self.host}:{self.port}")
                sock.connect((self.host, self.port))
                print(f"送信メッセージ: {message}")
                sock.sendall(packet)
                print("送信完了")
            
            return True
        except Exception as e:
            print(f"TCP/IP通信での送信に失敗: {e}")
            return False

File: api/bouyomi_client.cpp
#include "bouyomi_client.h"
#include <QUrl>
#include <QUrlQuery>
#include <QNetworkRequest>
#include <QNetworkReply>
#include <QDataStream>
#include <QDebug>

BouyomiClient::BouyomiClient(const QString& host, int port, int httpPort, QObject *parent)
    : QObject(parent), m_host(host), m_port(port), m_httpPort(httpPort)
{
    m_networkManager = new QNetworkAccessManager(this);
}

bool BouyomiClient::speak(const QString& message, int speed, int tone, int volume, int voice)
{
    // まずHTTP通信を試す
    if (speakHttp(message, speed, tone, volume, voice)) {
        return true;
    }
    
    // HTTPが失敗した場合はTCPを試す
    return speakTcp(message, speed, tone, volume, voice);
}

bool BouyomiClient::speakHttp(const QString& message, int speed, int tone, int volume, int voice)
{
    QUrl url(QString("http://%1:%2/talk").arg(m_host).arg(m_httpPort));
    QUrlQuery query;
    query.addQueryItem("text", message);
    if (voice != -1) query.addQueryItem("voice", QString::number(voice));
    if (volume != -1) query.addQueryItem("volume", QString::number(volume));
    if (speed != -1) query.addQueryItem("speed", QString::number(speed));
    if (tone != -1) query.addQueryItem("tone", QString::number(tone));
    
    url.setQuery(query);
    
    QNetworkRequest request(url);
    QNetworkReply* reply = m_networkManager->get(request);
    
    connect(reply, &QNetworkReply::finished, [this, reply, message]() {
        bool success = (reply->error() == QNetworkReply::NoError);
        emit speechFinished(success, message);
        reply->deleteLater();
    });
    
    return true; // 非同期のため、リクエスト開始時点でtrueを返す
}

bool BouyomiClient::speakTcp(const QString& message, int speed, int tone, int volume, int voice)
{
    QTcpSocket* socket = new QTcpSocket(this);
    socket->connectToHost(m_host, m_port);
    
    if (!socket->waitForConnected(3000)) {
        qDebug() << "TCP Connection failed:" << socket->errorString();
        socket->deleteLater();
        return false;
    }
    
    QByteArray messageBytes = message.toUtf8();
    QByteArray packet;
    QDataStream stream(&packet, QIODevice::WriteOnly);
    stream.setByteOrder(QDataStream::LittleEndian);
    
    // パケット構造を構築 (struct.pack('<HhhhhBi', ...))
    stream << (quint16)0x0001;      // Command
    stream << (qint16)speed;        // Speed
    stream << (qint16)tone;         // Tone
    stream << (qint16)volume;       // Volume
    stream << (qint16)voice;        // Voice
    stream << (quint8)0;            // Encoding (UTF-8)
    stream << (qint32)messageBytes.length();
    packet.append(messageBytes);
    
    socket->write(packet);
    socket->waitForBytesWritten(3000);
    socket->disconnectFromHost();
    
    connect(socket, &QTcpSocket::disconnected, socket, &QObject::deleteLater);
    
    return true;
}
